cluster/nodes.go

```diff
@@ -0,0 +1,228 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cluster
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/lib/pq"
+
+	uuid "github.com/satori/go.uuid"
+)
+
+type StorageGroupNode struct {
+	StorageGroupID *uuid.UUID
+	Num            int32
+	Node           *Node
+}
+
+type Node struct {
+	ID       uuid.UUID
+	Name     string
+	K8sLabel string
+	Volumes  []*Volume
+}
+
+func NewNode(name, k8sLabel string) (*Node, error) {
+	// validate node_name, must follow linux host naming rules
+	if len(name) > 253 {
+		return nil, errors.New("host name cannot be longer than 253 characters")
+	}
+	hostNameParts := strings.Split(name, ".")
+	// only alpha numerical with dashes segments are allowed
+	var re = regexp.MustCompile(`^[a-z0-9-]{1,63}$`)
+	for _, part := range hostNameParts {
+		if !re.MatchString(part) {
+			errMsg := fmt.Sprintf("invalid node name segment: `%s`", part)
+			return nil, errors.New(errMsg)
+		}
+	}
+	// validate kubernetes node label according to documentation:
+	// https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/
+	if len(k8sLabel) > 63 {
+		return nil, errors.New("kubernetes label cannot be longer than 63 characters")
+	}
+	var k8sre = regexp.MustCompile(`^[a-zA-Z0-9][a-z0-9-_.]{0,61}[a-zA-Z0-9]$`)
+	if !k8sre.MatchString(k8sLabel) {
+		errMsg := fmt.Sprintf("invalid kubernetes label: `%s`", k8sLabel)
+		return nil, errors.New(errMsg)
+	}
+	return &Node{ID: uuid.NewV4(), Name: name, K8sLabel: k8sLabel}, nil
+}
+
+// NodeAdd adds a new node for the cluster to administer
+func NodeAdd(ctx *Context, name, k8sLabel string) (*Node, error) {
+	// if we can instantiate a new node from the data, insert it
+	node, err := NewNode(name, k8sLabel)
+	if err != nil {
+		return nil, err
+	}
+	// insert the node in the DB
+	query := `INSERT INTO
+				nodes ("id", "name", "k8s_label", "sys_created_by")
+			  VALUES
+				($1, $2, $3, $4)`
+	tx, err := ctx.MainTx()
+	if err != nil {
+		return nil, err
+	}
+	// Execute query
+	_, err = tx.Exec(query, node.ID, node.Name, node.K8sLabel, ctx.WhoAmI)
+	if err != nil {
+		return nil, err
+	}
+	return node, nil
+}
+
+func GetNodeByName(ctx *Context, name string) (*Node, error) {
+	// Get user from tenants database
+	query := `
+		SELECT 
+				n.id,n.name,n.k8s_label
+			FROM 
+				nodes n
+			WHERE n.name=$1 LIMIT 1`
+
+	row := GetInstance().Db.QueryRow(query, name)
+	node := Node{}
+	err := row.Scan(&node.ID, &node.Name, &node.K8sLabel)
+	if err != nil {
+		return nil, err
+	}
+
+	return &node, nil
+}
+
+type Volume struct {
```

Comment:
can we rename this to NodeVolume so that it is consistent with the node_volumes table -Cesar